File: src/engine.rs
// src/engine.rs

use crate::ast::{Expr, Op, Statement};
use std::collections::HashMap;
use std::error::Error;
use std::fs::File;

/// Eksekusi query dan kembalikan hasil sebagai (headers, rows)
pub fn execute_query(query: Statement) -> Result<(Vec<String>, Vec<Vec<String>>), Box<dyn Error>> {
    match query {
        Statement::Select {
            columns,
            table,
            where_clause,
            limit,
        } => {
            // 1. Buka File CSV
            let file =
                File::open(&table).map_err(|_| format!("Gagal membuka file: {}", table))?;
            let mut rdr = csv::Reader::from_reader(file);

            // 2. Baca Header (Nama Kolom)
            let headers = rdr.headers()?.clone();
            let mut col_map = HashMap::new();
            for (i, h) in headers.iter().enumerate() {
                col_map.insert(h.to_string(), i);
            }

            // 3. Tentukan kolom output
            let output_headers: Vec<String> = if columns.len() == 1 && columns[0] == "*" {
                headers.iter().map(|s| s.to_string()).collect()
            } else {
                columns.clone()
            };

            // 4. Loop Data Baris per Baris
            let mut count = 0;
            let mut results: Vec<Vec<String>> = Vec::new();

            for result in rdr.records() {
                let record = result?;

                // 5. Cek WHERE (Filter)
                if let Some(ref expr) = where_clause {
                    if !eval(expr, &record, &col_map) {
                        continue;
                    }
                }

                // 6. Ambil Data Terpilih
                let mut row_output = Vec::new();
                for col_name in &columns {
                    if col_name == "*" {
                        for field in &record {
                            row_output.push(field.to_string());
                        }
                    } else if let Some(&idx) = col_map.get(col_name) {
                        if let Some(val) = record.get(idx) {
                            row_output.push(val.to_string());
                        }
                    }
                }
                results.push(row_output);

                // 7. Cek LIMIT
                count += 1;
                if let Some(max) = limit {
                    if count >= max {
                        break;
                    }
                }
            }

            Ok((output_headers, results))
        }
    }
}

fn eval(expr: &Expr, row: &csv::StringRecord, col_map: &HashMap<String, usize>) -> bool {
    match expr {
        Expr::BinaryOp { left, op, right } => {
            match op {
                // KELOMPOK 1: LOGIKA (AND / OR)
                Op::And => {
                    let left_is_correct = eval(left, row, col_map);
                    let right_is_correct = eval(right, row, col_map);
                    left_is_correct && right_is_correct
                }
                Op::Or => {
                    let left_is_correct = eval(left, row, col_map);
                    let right_is_correct = eval(right, row, col_map);
                    left_is_correct || right_is_correct
                }

                // KELOMPOK 2: PERBANDINGAN STRING (untuk Identifier vs StringLiteral)
                Op::Equal | Op::NotEqual => {
                    // Cek apakah ini perbandingan string
                    if let (Some(left_str), Some(right_str)) =
                        (get_string_value(left, row, col_map), get_string_value(right, row, col_map))
                    {
                        match op {
                            Op::Equal => left_str == right_str,
                            Op::NotEqual => left_str != right_str,
                            _ => false,
                        }
                    } else {
                        // Fallback ke perbandingan numerik
                        let left_val = get_value(left, row, col_map);
                        let right_val = get_value(right, row, col_map);
                        match op {
                            Op::Equal => (left_val - right_val).abs() < f64::EPSILON,
                            Op::NotEqual => (left_val - right_val).abs() > f64::EPSILON,
                            _ => false,
                        }
                    }
                }

                // KELOMPOK 3: PERBANDINGAN NUMERIK
                _ => {
                    let left_val = get_value(left, row, col_map);
                    let right_val = get_value(right, row, col_map);

                    match op {
                        Op::GreaterThan => left_val > right_val,
                        Op::LessThan => left_val < right_val,
                        Op::GreaterThanOrEq => left_val >= right_val,
                        Op::LessThanOrEq => left_val <= right_val,
                        _ => false,
                    }
                }
            }
        }
        _ => false,
    }
}

// Helper untuk mengambil nilai string dari ekspresi
fn get_string_value(expr: &Expr, row: &csv::StringRecord, col_map: &HashMap<String, usize>) -> Option<String> {
    match expr {
        Expr::StringLiteral(s) => Some(s.clone()),
        Expr::Literal(s) => Some(s.clone()),
        Expr::Identifier(col_name) => {
            if let Some(&idx) = col_map.get(col_name) {
                row.get(idx).map(|s| s.to_string())
            } else {
                None
            }
        }
        _ => None,
    }
}

// Helper untuk mengambil nilai numerik dari ekspresi
fn get_value(expr: &Expr, row: &csv::StringRecord, col_map: &HashMap<String, usize>) -> f64 {
    match expr {
        Expr::Number(n) => *n,
        Expr::Identifier(col_name) => {
            if let Some(&idx) = col_map.get(col_name) {
                if let Some(val_str) = row.get(idx) {
                    return val_str.parse::<f64>().unwrap_or(0.0);
                }
            }
            0.0
        }
        _ => 0.0,
    }
}

File: src/engine.py
"""
engine.py - Query Execution Engine untuk CSV_QL

═══════════════════════════════════════════════════════════════════════════════
                        📋 INSTRUKSI PENGISIAN
═══════════════════════════════════════════════════════════════════════════════

Modul ini mengeksekusi query terhadap file CSV dan mengembalikan hasil.
Engine adalah tahap terakhir dari pipeline kompilasi.

TUGAS YANG HARUS DIKERJAKAN:
-----------------------------

1. Implementasi fungsi execute_query(query: Statement) -> Tuple[List[str], List[List[str]]]
   yang:
   - Membuka file CSV
   - Membaca header (nama kolom)
   - Membaca data baris per baris
   - Mengevaluasi WHERE clause (jika ada)
   - Memilih kolom yang diminta
   - Membatasi hasil sesuai LIMIT (jika ada)
   - Mengembalikan tuple (headers, rows)

2. Implementasi fungsi eval(expr: Expr, row: dict) -> bool
   yang mengevaluasi expression dengan data dari baris CSV:
   - BinaryOp dengan AND -> eval kiri AND eval kanan
   - BinaryOp dengan OR -> eval kiri OR eval kanan
   - BinaryOp dengan perbandingan (=, !=, >, <, >=, <=)
     - Untuk string: bandingkan sebagai string
     - Untuk angka: convert dan bandingkan

3. Implementasi helper functions:
   - get_string_value(expr, row) -> Optional[str]
     Ambil nilai string dari expression
   - get_value(expr, row) -> float
     Ambil nilai numerik dari expression

REFERENSI:
----------
Lihat file Rust: src/engine.rs

TIPS:
-----
- Gunakan module 'csv' bawaan Python: csv.DictReader untuk baca CSV
- DictReader otomatis mapping kolom ke nama
- Perhatikan tipe data: string vs angka
- Untuk perbandingan float, gunakan threshold kecil (epsilon)

═══════════════════════════════════════════════════════════════════════════════
"""

import csv
from typing import Tuple, List, Dict, Optional
from ast_nodes import (Statement, Expr, Op, BinaryOp, Literal, 
                       StringLiteral, Number, Identifier, SelectStatement)


def execute_query(query) -> Tuple[List[str], List[List[str]]]:  # query: Statement
    """
    Eksekusi query dan kembalikan hasil.
    
    Args:
        query: Statement AST dari parser
        
    Returns:
        Tuple berisi (headers, rows)
        - headers: List nama kolom
        - rows: List baris data (setiap baris adalah list string)
        
    Raises:
        Exception: Jika ada error saat eksekusi (file tidak ada, dll)
    """
    
    # TODO: Implementasi
    #
    # 1. Buka file CSV
    with open(query.table, 'r', newline='', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        
        # 2. Dapatkan header
        all_headers = list(reader.fieldnames) if reader.fieldnames else []
        
        # 3. Tentukan output headers
        if query.columns == ["*"]:
            output_headers = all_headers
        else:
            output_headers = query.columns
        
        # 4. Proses setiap baris
        results = []
        count = 0
        
        for row in reader:
            # 5. Evaluasi WHERE clause
            if query.where_clause:
                if not eval_expr(query.where_clause, row):
                    continue
            
            # 6. Ambil kolom yang diminta
            if query.columns == ["*"]:
                row_data = [row[col] for col in all_headers]
            else:
                row_data = [row.get(col, "") for col in query.columns]
            
            results.append(row_data)
            
            # 7. Cek LIMIT
            count += 1
            if query.limit and count >= query.limit:
                break
        
        return (output_headers, results)


def eval_expr(expr: Expr, row: Dict[str, str]) -> bool:
    """
    Evaluasi expression dengan data baris.
    
    Args:
        expr: Expression dari WHERE clause
        row: Dictionary data baris (nama kolom -> nilai)
        
    Returns:
        True jika baris memenuhi kondisi, False jika tidak
    """
    
    if isinstance(expr, BinaryOp):
        # LOGIKA (AND / OR)
        if expr.op == Op.AND:
            return eval_expr(expr.left, row) and eval_expr(expr.right, row)
        elif expr.op == Op.OR:
            return eval_expr(expr.left, row) or eval_expr(expr.right, row)
        
        # PERBANDINGAN STRING
        elif expr.op in (Op.EQUAL, Op.NOT_EQUAL):
            left_str = get_string_value(expr.left, row)
            right_str = get_string_value(expr.right, row)
            
            if left_str is not None and right_str is not None:
                if expr.op == Op.EQUAL:
                    return left_str == right_str
                else:
                    return left_str != right_str
            
            # Fallback ke numerik
            left_val = get_value(expr.left, row)
            right_val = get_value(expr.right, row)
            
            if expr.op == Op.EQUAL:
                return abs(left_val - right_val) < 1e-9
            else:
                return abs(left_val - right_val) > 1e-9
        
        # PERBANDINGAN NUMERIK
        else:
            left_val = get_value(expr.left, row)
            right_val = get_value(expr.right, row)
            
            if expr.op == Op.GREATER_THAN:
                return left_val > right_val
            elif expr.op == Op.LESS_THAN:
                return left_val < right_val
            elif expr.op == Op.GREATER_THAN_OR_EQ:
                return left_val >= right_val
            elif expr.op == Op.LESS_THAN_OR_EQ:
                return left_val <= right_val
    
    return False


def get_string_value(expr: Expr, row: Dict[str, str]) -> Optional[str]:
    """
    Ambil nilai string dari expression.
    
    Args:
        expr: Expression (Identifier, StringLiteral, atau Literal)
        row: Dictionary data baris
        
    Returns:
        String value, atau None jika tidak bisa diambil
    """
    
    if isinstance(expr, StringLiteral):
        return expr.value
    elif isinstance(expr, Literal):
        return expr.value
    elif isinstance(expr, Identifier):
        return row.get(expr.name)
    
    return None


def get_value(expr: Expr, row: Dict[str, str]) -> float:
    """
    Ambil nilai numerik dari expression.
    
    Args:
        expr: Expression (Number atau Identifier)
        row: Dictionary data baris
        
    Returns:
        Float value (0.0 jika tidak bisa dikonversi)
    """
    
    if isinstance(expr, Number):
        return expr.value
    elif isinstance(expr, Identifier):
        val_str = row.get(expr.name, "0")
        try:
            return float(val_str)
        except ValueError:
            return 0.0
    
    return 0.0
